### PR TITLE
dev-python/PyQt5 and friends: bumps and python3.11

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
This bumps some packages to the latest version and adds python 3.11 compatibility to all.

The patch is from upstream: https://www.riverbankcomputing.com/hg/sip/rev/3d656f0765f508ea9714b5e7c9ce00e304e6f600
And is required due to the following change in [python 3.11](https://docs.python.org/3.11/whatsnew/3.11.html):
> Global inline flags (e.g. (?i)) can now only be used at the start of the regular expressions. Using them not at the start
>  of expression was deprecated since Python 3.6. (Contributed by Serhiy Storchaka in [bpo-47066](https://bugs.python.org/issue?@action=redirect&bpo=47066).)

CC: @gentoo/qt

### CODE REVIEW

**Comment by AndrewAmmerlaan** [comment]
Update: Also added python 3.11 to dev-python/QtPy. Unfortunately pyside2 currently does not compile with python 3.11. Upstream development has mostly moved to pyside6 for Qt6, so I don't expect this to get fixed. Python 3.10 on pyside2 already required a couple of downstream patches. Instead I re-apply the same workaround we used before with python 3.10: if python 3.11 is enabled we hard depend on PyQt5 instead of allowing the user to choose which Qt4Python implementation to use.

**Comment by AndrewAmmerlaan** [comment]
> Update: Also added python 3.11 to dev-python/QtPy. Unfortunately pyside2 currently does not compile with python 3.11. Upstream development has mostly moved to pyside6 for Qt6, so I don't expect this to get fixed. Python 3.10 on pyside2 already required a couple of downstream patches. Instead I re-apply the same workaround we used before with python 3.10: if python 3.11 is enabled we hard depend on PyQt5 instead of allowing the user to choose which Qt4Python implementation to use.

Update2: Added python 3.11 to pytest-qt with the same workaround.

@gentoo/qt A _huge_ part of the scientific python software stack depends on this, please let me know if it is okay to merge this so I can continue testing the python packages higher up in the dependency tree with python 3.11.

File: dev-python/sip/sip-6.6.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+RDEPEND="
+	!<dev-python/sip-4.19.25-r1[${PYTHON_USEDEP}]
+	!=dev-python/sip-5.5.0-r0[${PYTHON_USEDEP}]
+	dev-python/packaging[${PYTHON_USEDEP}]
+	dev-python/toml[${PYTHON_USEDEP}]
+"
+
+PATCHES=(
+	"${FILESDIR}/${P}-python311-invalid-regex.patch"
```
Comment (Pesa) [inline dev-python/sip/sip-6.6.1.ebuild:33]: you seem to have dropped the existing patch, but I believe it's still necessary

**Review by Pesa — CHANGES_REQUESTED** [review]
I was hoping for a quick sip-6.6.2 upstream release since there are a large number of unreleased fixes, but well... if this is blocking other work then I guess we can proceed, it's ~arch anyway.

There seems to be another issue with py3.11 though https://www.riverbankcomputing.com/pipermail/pyqt/2022-May/044668.html

**Comment by Pesa** [comment]
Also note that I haven't tested this locally and I won't be able to do that for at least another week or two (hardware issues).

**Review by Pesa — CHANGES_REQUESTED** [review]

File: dev-python/sip/sip-6.6.1.ebuild (hunk 1 above)
Comment (AndrewAmmerlaan) [reply dev-python/sip/sip-6.6.1.ebuild:34]: Thanks, I had missed this one because apparently my tree was too old when I started working on this.

**Comment by AndrewAmmerlaan** [comment]
> There seems to be another issue with py3.11 though https://www.riverbankcomputing.com/pipermail/pyqt/2022-May/044668.html

That's unfortunate, there doesn't seem to be a patch for this yet. As I understand it, this will only surface in some applications. Perhaps we can proceed with adding python 3.11 here and then we'll find out which PyQt applications are affected and which ones aren't as we test them with python 3.11. We'll just have to wait with adding python 3.11 compatibility to those packages that are affected by this issue.

**Comment by ionenwks** [comment]
On a side-note, PyQt6 is already prepared, will be adding it after qt6 is in tree which should happen soon'ish (just mentioning not to risk duplicating work, see old https://github.com/gentoo/qt/pull/253 -- it also use sip pep517 which I guess PyQt5 should be able to use too :eyes:)

**Comment by AndrewAmmerlaan** [comment]
> On a side-note, PyQt6 is already prepared, will be adding it after qt6 is in tree which should happen soon'ish (just mentioning not to risk duplicating work, see old [gentoo/qt#253](https://github.com/gentoo/qt/pull/253) -- it also use sip pep517 which I guess PyQt5 should be able to use too eyes)

I think we are going to want to add python 3.11 to both PyQt5 and PyQt6 anyway, since there are probably going to be many applications that aren't ready yet for Qt6 but might still work with python 3.11.

**Comment by thesamesam** [comment]
@pesa unfortunately this ends up blocking a fair bit (e.g. portage -> selinux -> pyqt5) Are you alright with proceeding on the basis @AndrewAmmerlaan mentions above? (Just being careful that rdeps actually work), or do you want to wait?

**Review by Pesa — COMMENTED** [review]
Yep, sorry, the delay is simply due to me forgetting about this PR, not because I want to wait for the next sip release. Thanks for reminding me!

I have just one more comment on the sip ebuild. The PyQt-related stuff LGTM.

**Comment by thesamesam** [comment]
Thanks!

FWIW: I looked at veusz for a bug (don't use it myself, but wanted to give pointers in the right direction, so poked at it) and w/ new sip, it looks like some minor adjustments were needed: https://bugs.gentoo.org/823005#c10. Doesn't bother me but just a heads up in case other stuff breaks. It looks like veusz might just be the odd one out here, or at least this will slowly be a problem of the past w/ PEP517?

File: dev-python/sip/sip-6.6.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# Copyright 1999-2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_{8..11} )
+DISTUTILS_USE_PEP517=setuptools
```
[the remark this answers is not in the document]
Comment (Pesa) [reply dev-python/sip/sip-6.6.1.ebuild:7]: ok, then it needs to be added

[the remark this answers is not in the document]
Comment (AndrewAmmerlaan) [reply dev-python/sip/sip-6.6.1.ebuild:7]: > so how is the setuptools rdepend handled by the new PEP517 mode? Note that sip uses setuptools at runtime

This is a good point, hadn't thought of that. I added the setuptools rdep manually.

**Comment by Pesa** [comment]
@AndrewAmmerlaan Did you push an updated version already? I don't see setuptools in the sip ebuild.

**Comment by AndrewAmmerlaan** [comment]
> @AndrewAmmerlaan Did you push an updated version already? I don't see setuptools in the sip ebuild.

I did, but I see I made a mistake in the rebase and it ended up in the wrong commit. Will fix it in a bit.

**Comment by AndrewAmmerlaan** [comment]
> > @AndrewAmmerlaan Did you push an updated version already? I don't see setuptools in the sip ebuild.
> 
> I did, but I see I made a mistake in the rebase and it ended up in the wrong commit. Will fix it in a bit.

This is what I get for doing late-night rebasing. It should be correct now.

**Comment by ionenwks** [comment]
> There seems to be another issue with py3.11 though https://www.riverbankcomputing.com/pipermail/pyqt/2022-May/044668.html

About that one, was just trying and it seems fixed if using python 3.11 beta3 (can only reproduce if I downgrade to beta1)

**Pull request closed without merging**